### DISCUSSION TITLE
Torch and numpy interoperability?

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [xamarin/TorchSharp#394] [Ideas]
Is there a plan to support [Torch and numpy interoperability](https://medium.com/axinc-ai/conversion-between-torch-and-numpy-operators-ce189b3882b1) in .NET?

>Numpy is a popular open source library used for mathematical and scientific computing in Python. It enables efficient operations on large multi-dimensional arrays, and has a large ecosystem of supporting libraries:

>Instead of reinventing the wheel, PyTorch interoperates really well with Numpy to leverage its existing ecosystem of tools and libraries.

Currently, users of Tensorflow.NET use either
-  [Numpy.NET](https://github.com/SciSharp/Numpy.NET) or 
- [NumSharp](https://github.com/SciSharp/NumSharp) or 
- [NumSharp.Lite](https://github.com/SciSharp/NumSharp.Lite)
- {Non SciSharp} [numpy.net](https://github.com/Quansight-Labs/numpy.net) or [NumpyDotNet@nuget](https://www.nuget.org/packages/NumpyDotNet/)

The .NET community need Microsoft to support the community to (perhaps) coordinate/prioritize the different efforts of bringing Numpy to .NET.

**Comment by aiqinxuancai** [comment]
When translating python scripts, I often encounter the need for libraries such as sklearn, and currently I can only go to the reference py implementation for rewriting, which is really tiring.

**Reply by GeorgeS2019** [reply]
This issue has been discussed many months before. 

The key is to gather community effort and **reusability**. If someone has ported some python code e.g. from sklearn to .NET, we need to think of reusability on top of existing framework e.g. ML.NET. (this is just my opinion, we need a further discussion on that :-))

We need a **concerted community effort** to plan how to move forwards so we have .NET substitutions for e.g. sklearn Spacy etc. 

The more developers like yourself feedback this, the more the dotnet team will take this into consideration.

Keep providing feedback.

Thanks for keep contributing!

**Reply by filipeclduarte** [reply]
Which numerical library above mentioned should be used to port the Sklearn library to .NET and maintain interoperability with TorchSharp?

Also, as another route, I think it is possible to reborn Accord.NET, which is now archived. It is simple to Sklearn with the peculiarity of C#.

**Comment by GeorgeS2019** [comment]
TorchSharp's active community has grown significantly since the same time last year.  What seems inconceivable which is daily with python is now within reach using dotnet thanks to Torchsharp.

YES, we need more discussions.

For [Accord.NET its license ](https://github.com/accord-net/framework/blob/development/LICENSE) needs to be made available as MIT license.  Someone needs to persuade this transition since it is now a public archive. 

We need to discuss different strategies for how to do that.

One of the key reasons why TorchSharp is catching growing traction, in my view, was the decision and active commitment to making the portability from PyTorch to dotnet simpler AND keeping Torchsharp up-to-date with abundant and high coverage tests to demonstrate how to do the portability.

We need to learn from this success, to make machine learning for dotnet a more complete experience.

**Reply by filipeclduarte** [reply]
I understand what you have said and think those advances made really awesome. It is essential to continue discussing and elaborate a plan to build a fully interoperable companion of libraries for data manipulation, numerical computing in CPU (for small matrices and interoperable with TorchSharp), probability, statistics, general machine learning, data visualization, etc. 

Right now, is there any .NET numerical computing library that is interoperable with TorchSharp?

**Reply by GeorgeS2019** [reply]
https://en.m.wikipedia.org/wiki/List_of_numerical_libraries

Can u be more specific?

The discussion need to be e.g. for pytotch, what are the typical options, are these available in .NET?  If so, are they  interoperable with TorchSharp?

**Reply by filipeclduarte** [reply]
I mean ... Do any of those libraries mentioned in the first post have integration with TorchSharp? Considering we are talking about the library on the .NET platform. Maybe some method like `tensor.toNumpy()` conversion? 

From what I've seen, [NumpyDotNet](https://github.com/Quansight-Labs/numpy.net) is the most compatible with the NumPy API.

**Reply by GeorgeS2019** [reply]
I do not speak for the main TorchSharp developers from Microsoft.

I hope, if there are sufficient users here who have similar wishes for  [`torch.from_numpy`](https://pytorch.org/docs/stable/generated/torch.from_numpy.html?msclkid=1ef085cfc5ab11ec8673a89e26c1929a) and/or [PyTorch Tensor to Numpy](https://pythonguides.com/pytorch-tensor-to-numpy/?msclkid=2f75b55dc5ad11ecbc54c8bf626a2634),  Perhaps the wishes will be considered.

**Reply by GeorgeS2019** [reply]
@filipeclduarte  => [FYI](https://github.com/Quansight-Labs/numpy.net/issues/27)

**Comment by GeorgeS2019** [comment]
@kaiidams 

In your most recent [TorchAudio example using TorchSharp](https://github.com/dotnet/TorchSharp/pull/649): [SPEECH COMMAND CLASSIFICATION WITH TORCHAUDIO](https://pytorch.org/tutorials/intermediate/speech_command_classification_with_torchaudio_tutorial.html)

There are a few Pytorch Numpy interoperability statements
```python
waveform.t().numpy()
waveform_first.numpy()
```

Could you please share your views on how would TorchSharp community approach these interoperability statements? Primarily based on the discussion above and the availability of different Numpy for .NET libraries.

CC @NiklasGustafsson

**Comment by GeorgeS2019** [comment]
[Related discussion](https://github.com/dotnet/TorchSharp/pull/680#issuecomment-1205657199)

**Comment by NiklasGustafsson** [comment]
>> Is there a plan to support Torch and numpy interoperability in .NET?

As far as TorchSharp is concerned, the answer is no. We're trying to keep the scope very narrow and clean, so that we can manage it without too many dependencies and with minimal staff. I also would like to see a de facto standard .NET numerics library, but we will not be using TorchSharp as a forcing function.

**Comment by GeorgeS2019** [comment]
@NiklasGustafsson 

Now that we have tensor.from_array(),   what is your recommended way to do tensor.to_array

Typically, in pytorch, this is tensor.numpy()

Any suggestion?

Ideally there is **a small documentation** to guide how best to do` tensor.from_array` and `tensor.to_array` in terms of performance and best align with .NET best practice.

**Comment by NiklasGustafsson** [comment]
Tensor.data\<T\> is what you're looking for.

**Comment by GeorgeS2019** [comment]


**Reply by NiklasGustafsson** [reply]
Tensor.data<T> is used in 158 unit tests... :-)

**Comment by GeorgeS2019** [comment]
FYI: 
### [ Bridge with NumPy](https://pytorch.org/tutorials/beginner/blitz/tensor_tutorial.html#bridge-with-numpy)

A short documentation section on **.NET arrays and tensors.**?
- System.Array to Tensor
- Tensor to System.Array

**Reply by NiklasGustafsson** [reply]
Let's not describe from_array and data<T> as having anything to do with numpy. They do not -- they bridge between .NET arrays and tensors. As I have pointed out elsewhere, numpy is a Python library that TorchSharp has no plans to integrate with.

**Reply by GeorgeS2019** [reply]
Agree,!

What u mean by this?
>  I also would like to see a de facto standard .NET numerics library,

**Reply by NiklasGustafsson** [reply]
That it would be great if there were something in the .NET world that was like numpy in capabilities and broad acceptance / deployment. I am not aware of any such library existing.
